### PR TITLE
Mats/CERES-428 Created a GET /api/search/lookup api aligning with OpenTsdb API.

### PR DESCRIPTION
**Pull request by matsboman**
# Resolves

https://jira.rax.io/browse/CERES-428

# What

We need to create a GET /api/search/lookup api aligning with OpenTsdb api.

# How

Implement support for querying for tagKeys and tagValues according to the tsdb specification.
http://opentsdb.net/docs/build/html/api_http/search/lookup.html

## How to test

Execute the tests in LookupApiControllerTest class.

# Why

We need this for grafana to be able to call our endpoint.

# TODO

N/A

### CODE REVIEW

File: src/main/java/com/rackspace/ceres/app/model/MetricNameAndMultiTags.java [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+
+import java.util.List;
+import java.util.Map;
+
+import lombok.Data;
+
+@Data
+public class MetricNameAndMultiTags {
+    String metricName;
+    List<Map<String, String>> tags;
```
Comment (matsboman) [inline src/main/java/com/rackspace/ceres/app/model/MetricNameAndMultiTags.java:11]: We need to support multiple entries with same key. It is easier implementation and easier to understand.
e.g
host=h-1,
host=h-2,
host=*

Comment (harageth) [reply src/main/java/com/rackspace/ceres/app/model/MetricNameAndMultiTags.java:11]: Is each map storing more than one value? Could we make it a list of Pair's? https://docs.oracle.com/javase/9/docs/api/javafx/util/Pair.html

File: src/main/java/com/rackspace/ceres/app/model/SeriesData.java [hunk 2]
```diff
@@ -0,0 +1,10 @@
+package com.rackspace.ceres.app.model;
+
+import java.util.Map;
+
+import lombok.Data;
+
+@Data
+public class SeriesData {
+    Map<String, String> tags;
```
Comment (harageth) [inline src/main/java/com/rackspace/ceres/app/model/SeriesData.java:9]: Shouldn't we have the metrics in here with the tags?

File: src/main/java/com/rackspace/ceres/app/web/TsdbQueryController.java [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+
+    @GetMapping("/lookup")
+    public Mono<LookupResult> query(@RequestParam(name = "m") String m,
+                                    @RequestParam(name = "limit") Integer limit,
+                                    @RequestHeader(value = "X-Tenant", required = true) String tenantHeader) {
+
+        List<SeriesData> results = new ArrayList<SeriesData>();
+        final MetricNameAndMultiTags metricNameAndTags = getMetric(m);
+
+        return metadataService.getTagKeys(tenantHeader, metricNameAndTags.getMetricName())
```
Comment (harageth) [inline src/main/java/com/rackspace/ceres/app/web/TsdbQueryController.java:37]: Can we create some flux utility functions to do most of this? Perhaps that should be a refactoring ticket all on its own that gets pulled into next sprint.

File: src/main/java/com/rackspace/ceres/app/model/SeriesData.java (hunk 2 above)
Comment (matsboman) [reply src/main/java/com/rackspace/ceres/app/model/SeriesData.java:9]: This is a wrapper class so each tag comes out as an object in the reply.

File: src/main/java/com/rackspace/ceres/app/model/MetricNameAndMultiTags.java (hunk 1 above)
Comment (matsboman) [reply src/main/java/com/rackspace/ceres/app/model/MetricNameAndMultiTags.java:11]: A bit reluctant to create a dependency on oracle. I think map works ok.

File: src/main/java/com/rackspace/ceres/app/web/TsdbQueryController.java (hunk 3 above)
Comment (matsboman) [reply src/main/java/com/rackspace/ceres/app/web/TsdbQueryController.java:37]: Sounds like a good idea. Let's write up a JIRA.

File: src/main/java/com/rackspace/ceres/app/model/MetricNameAndMultiTags.java (hunk 1 above)
Comment (harageth) [reply src/main/java/com/rackspace/ceres/app/model/MetricNameAndMultiTags.java:11]: Oh yeah didn't realize that when I linked it. Good call.

File: src/main/java/com/rackspace/ceres/app/web/TsdbQueryController.java [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+                    }
+                }
+            });
+        }
+    }
+
+    private MetricNameAndMultiTags getMetric(String m) {
+        MetricNameAndMultiTags metricNameAndTags = new MetricNameAndMultiTags();
+        List<Map<String, String>> tags = new ArrayList<Map<String, String>>();
+        if (m.matches(".*\\{.*\\}$")) {
```
Comment (rohitsngh27) [inline src/main/java/com/rackspace/ceres/app/web/TsdbQueryController.java:75]: can we create a string constant for this pattern?

File: src/main/java/com/rackspace/ceres/app/web/TsdbQueryController.java [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+
+import java.util.*;
+
+/**
+ * Native Ceres search API endpoints.
+ */
+@RestController
+@RequestMapping("/api/search")
+@Profile("query")
+public class TsdbQueryController {
```
Comment (rohitsngh27) [inline src/main/java/com/rackspace/ceres/app/web/TsdbQueryController.java:21]: I think the file name should be `LookupApiController`

File: src/main/java/com/rackspace/ceres/app/web/TsdbQueryController.java (hunk 3 above)
Comment (rohitsngh27) [reply src/main/java/com/rackspace/ceres/app/web/TsdbQueryController.java:37]: and please include moving utility methods from controller class to a service class in that JIRA ticket

**Review by rohitsngh27 — CHANGES_REQUESTED** [review]
Just some minor comments.

**Review by rohitsngh27 — DISMISSED** [review]

**Review by rohitsngh27 — APPROVED** [review]

**Pull request merged**